libchisel/src/trimexports.rs

```diff
@@ -0,0 +1,201 @@
+            ],
+        }
+    }
+
+    /// Constructs a whitelist with the parity wasm export interface preset.
+    fn pwasm() -> Self {
+        ExportWhitelist {
+            entries: vec![ExportEntry::new("_call".to_string(), Internal::Function(0))],
+        }
+    }
+
+    /// Looks up a given export entry in the whitelist and returns true if it is valid.
+    fn lookup(&mut self, export: &ExportEntry) -> bool {
+        if let Some(thing) = self.entries.iter().find(|x| {
+            export.field() == x.field() && cmp_internal_variant(export.internal(), x.internal())
+        }) {
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl TrimExports {
+    /// Constructs an empty `trimexports` context.
+    fn new() -> Self {
+        TrimExports {
+            whitelist: ExportWhitelist::new(),
+        }
+    }
+
+    /// Takes a given preset string and constructs a context with the
+    /// corresponding whitelist.
+    fn with_preset(preset: &str) -> Self {
+        match preset {
+            "ewasm" => TrimExports {
+                whitelist: ExportWhitelist::ewasm(),
+            },
+            "pwasm" => TrimExports {
+                whitelist: ExportWhitelist::pwasm(),
+            },
+            _ => TrimExports::new(),
+        }
+    }
+
+    /// Iterates over the export section, if there is one, and removes
+    /// unnecessary entries.
+    fn trim_exports(&mut self, module: &mut Module) -> bool {
+        if let Some(section) = export_section_mut(module) {
+            let new_section = ExportSection::with_entries(
+                section
+                    .entries()
+                    .iter()
+                    .cloned()
+                    .filter(|x| self.whitelist.lookup(x))
+                    .collect(),
+            );
+
+            if new_section.entries().len() < section.entries().len() {
+                *section = new_section;
+                return true;
+            }
+
+            false
+        } else {
+            false
+        }
+    }
+}
+
+impl ModuleTranslator for TrimExports {
+    fn translate(mut self, module: &mut Module) -> Result<bool, String> {
+        Ok(self.trim_exports(module))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use parity_wasm::builder;
+
+    // Smoke tests
+    #[test]
+    fn builder_all_exports_good_ewasm() {
+        let mut module = builder::module()
+            .function()
+            .signature()
+            .build()
+            .body()
+            .build()
+            .build()
+            .export()
+            .field("main")
+            .internal()
+            .func(0)
+            .build()
+            .export()
+            .field("memory")
+            .internal()
+            .memory(0)
+            .build()
+            .build();
+
+        let trimmer = TrimExports::with_preset("ewasm");
+        let did_change = trimmer.translate(&mut module).unwrap();
+        assert!(!did_change);
```

Comment:
I think `assert_eq!(did_change, false)` is more readable here. The exclamation marks are too close.